tests/_ast/test_visitor.py

```diff
@@ -972,6 +972,44 @@ def h():
     assert v.refs == {"x"}
 
 
+def test_not_deleted_ref() -> None:
+    code = cleandoc(
+        """
+    x = 0
+    def fn():
+        if False:
+            z = x
+            del x
+    x
+    """
+    )
+
+    v = visitor.ScopedVisitor()
+    mod = ast.parse(code)
+    v.visit(mod)
+    assert v.defs == {"x", "fn"}
+    assert not v.refs
+    assert not v.deleted_refs
+
+
+def test_deleted_ref() -> None:
+    code = cleandoc(
+        """
+    def fn():
```

Comment:
I think this is only true when there's a global, otherwise del pushes the block to be in `scope`.

Maybe another bug when playing around but:

```python
x = 1
def f():
    global x
```

refs = {'x'}

but seems like refs = {}

but honestly, I wouldn't be opposed to just banning global

Reply:
> I think this is only true when there's a global, otherwise del pushes the block to be in scope.

I don't understand what you mean. This test passes; is there another test that should pass but doesn't?

Reply:
(Pushed a fix for the global test case.)

Reply:
The del case is the same as:

```python
x = 1
```

```python
def g():
    z = x
    x = 0
```

Which if you actually try to run gives you a NameError, because python- detecting the assign without global and scopes x.

We're find all the bugs because currently refs == {"x"} for this case - which is not really how this behaves, it should be refs == {}.

Likewise in the del case, I think refs == {}, deleted_refs == {} (since it's local effectively).

This changes when a global is present though.

```python
def g():
    global x
    z = x
    x = 0
```
refs == {"x"}

```python
def g():
    global x
    z = x
    del
```
refs == {"x"}, deleted_refs == {"x"}

Reply:
Oh got it! I see what you're saying. Thanks.